test/puppetlabs/jdbc_util/pool_test.clj

```diff
@@ -0,0 +1,110 @@
+                 :driver-class-name "org.postgresql.Driver"
+                 :jdbc-url "jdbc:postgresql:jdbc_util_test"
+                 :auto-commit false
+                 :connection-timeout 1200
+                 :idle-timeout 20002
+                 :max-lifetime 32001
+                 :connection-test-query "SELECT 1 FROM ANYWHERE;"
+                 :minimum-idle 5
+                 :maximum-pool-size 200
+                 ;; :metric-registry
+                 ;; :health-check-registry
+                 :pool-name "bob"})]
+    (testing "can set options on a HikariConfig"
+      (are [value getter] (= value (getter config))
+        "trogdor" .getUsername
+        "burn1n4t3" .getPassword
+        "org.postgresql.Driver" .getDriverClassName
+        "jdbc:postgresql:jdbc_util_test" .getJdbcUrl
+        false .isAutoCommit
+        1200 .getConnectionTimeout
+        20002 .getIdleTimeout
+        32001 .getMaxLifetime
+        "SELECT 1 FROM ANYWHERE;" .getConnectionTestQuery
+        5 .getMinimumIdle
+        200 .getMaximumPoolSize
+        "bob" .getPoolName))
+
+    (let [config (pool/options->hikari-config
+                  {:username "nobody"
+                   :password "noone"
+                   :data-source-class-name "org.postgresql.ds.PGSimpleDataSource"})]
+      (testing "can set data-source-class-name"
+        (is (= "org.postgresql.ds.PGSimpleDataSource"
+               (.getDataSourceClassName config)))))))
+
+(deftest delayed-init
+  (let [ready (atom false)
+        retries (atom 0)
+        mock-connection (reify
+                          Connection
+                          Closeable
+                          (close [_] nil))
+        mock-ds (reify
+                  DataSource
+                  (getConnection [_]
+                    (if @ready
+                      mock-connection
+                      (do
+                        (swap! retries inc)
+                        (Thread/sleep 500)
+                        (throw (SQLTransientConnectionException. "oops")))))
+                  (getConnection [this user pass]
+                    (.getConnection this)))]
+    (testing "the pool retries until it gets a connection"
+      (let [wrapped (pool/wrap-with-delayed-init mock-ds (fn [_] mock-ds) 1)]
+        (is (thrown? SQLTransientConnectionException
+                     (.getConnection (:datasource wrapped))))
+        (Thread/sleep 2000)
+        (swap! ready (constantly true))
+        (Thread/sleep 600)
+        (.getConnection (:datasource wrapped))
+        (is (<= 4 @retries 5))))  ; allow for some variance due to timing
+
+    (testing "if the init-fn throws an exception it doesn't give out connections"
+      (swap! ready (constantly true))  ; don't depend on state from the previous test
+      (let [wrapped (pool/wrap-with-delayed-init
+                     mock-ds (fn [_] (throw (RuntimeException.))) 1)]
+        (is (thrown? java.util.concurrent.ExecutionException
```

Comment:
So we should definitely have something that can get the root cause of the ExecutionException and log it. That's probably more appropriate to do in the consuming projects, but I feel like the docstring for wrap-with-delayed-init should explicitly call out that you should add that.